File: app/frontend/src/types/comparison.ts
export interface FeatureImportanceItem {
  category: string;
  weight: number;
}

export type ModelName = "global_model" | "your_model";
export type Confidence = "High" | "Medium" | "Low";

export interface ModelResult {
  name: ModelName;
  display_name: string;
  point_forecast: number[];
  p10: number[];
  p90: number[];
  mape: number;
  accuracy: number;
  confidence: Confidence;
  total_forecast: number;
  feature_importance: FeatureImportanceItem[] | null;
}

export interface ComparisonRequest {
  dataset_id: string;
  mapping: import("./dataset").ColumnMapping;
  horizon: number;
}

export interface ComparisonResponse {
  winner: ModelResult;
  alternative: ModelResult;
  winner_explanation: string;
  dates: string[];
  historical_dates: string[];
  historical_values: number[];
  backtest_holdout: number;
}

File: app/backend/tempolith/schemas/comparison.py
"""Comparison request and response schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .dataset import ColumnMapping
from .forecast import ForecastConfigIn


class FeatureImportanceItem(BaseModel):
    model_config = ConfigDict(frozen=True)

    category: str
    weight: float


class ModelResult(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    name: Literal["global_model", "your_model"]
    display_name: str
    point_forecast: list[float]
    p10: list[float]
    p90: list[float]
    mape: float
    accuracy: float
    confidence: Literal["High", "Medium", "Low"]
    total_forecast: float
    feature_importance: list[FeatureImportanceItem] | None = None


class ComparisonRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    dataset_id: str
    mapping: ColumnMapping
    horizon: int = Field(default=12, ge=1, le=256)
    forecast_config: ForecastConfigIn = Field(default_factory=ForecastConfigIn)


class ComparisonResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    winner: ModelResult
    alternative: ModelResult
    winner_explanation: str
    dates: list[str]
    historical_dates: list[str]
    historical_values: list[float]
    backtest_holdout: int
